src/credentials/DeviceAttestationVendorReserved.h

```diff
@@ -0,0 +1,218 @@
+};
+
+// extract elements out of the device attestation bytespan
+class DeviceAttestationVendorReservedDeconstructor
+{
+
+public:
+    DeviceAttestationVendorReservedDeconstructor() {}
+
+    // read TLV until first profile tag
+    CHIP_ERROR PrepareToReadVendorReservedElements(const ByteSpan & attestationElements, size_t count)
+    {
+        mNumVendorReservedData = count;
+        mAttestationData       = attestationElements;
+
+        tlvReader.Init(mAttestationData);
+        ReturnErrorOnFailure(tlvReader.Next(containerType, TLV::AnonymousTag));
+        ReturnErrorOnFailure(tlvReader.EnterContainer(containerType));
+
+        // position to first ProfileTag
+        while (true)
+        {
+            ReturnErrorOnFailure(tlvReader.Next());
+            if (!TLV::IsProfileTag(tlvReader.GetTag()))
+                break;
+        }
+        // positioned to first context tag (vendor reserved data)
+        mIsInitialized = true;
+        return CHIP_NO_ERROR;
+    }
+
+    CHIP_ERROR Next() { return tlvReader.Next(); }
+
+    size_t GetNumberOfElements() { return mNumVendorReservedData; }
+
+    /**
+     *  @brief Return next VendorReserved element. PrepareToReadVendorReservedElements must be called first.
+     *
+     *  @param[out] element  Next vendor Reserved element
+     *
+     *  @returns   CHIP_NO_ERROR on success
+     *             CHIP_ERROR_INCORRECT_STATE if PrepareToReadVendorReservedElements hasn't been called first
+     *             CHIP_END_OF_TLV if not further entries are present
+     */
+    CHIP_ERROR GetNextVendorReservedElement(struct VendorReservedElement & element)
+    {
+        VerifyOrReturnError(mIsInitialized, CHIP_ERROR_INCORRECT_STATE);
+
+        while (1)
+        {
+            ReturnErrorOnFailure(Next());
+            uint64_t tag = tlvReader.GetTag();
+            if (!TLV::IsProfileTag(tag))
+            {
+                continue;
+            }
+            // tag is profile tag
+            element.vendorId   = TLV::VendorIdFromTag(tag);
+            element.profileNum = TLV::ProfileNumFromTag(tag);
+            element.tagNum     = TLV::TagNumFromTag(tag);
+
+            return tlvReader.GetByteView(element.vendorReservedData);
+        }
+    }
+
+private:
+    size_t mNumVendorReservedData; // number of VendorReserved entries (could be 0)
+    ByteSpan mAttestationData;
+    bool mIsInitialized = false;
+    TLV::ContiguousBufferTLVReader tlvReader;
+    TLV::TLVType containerType = TLV::kTLVType_Structure;
+};
+
+class DeviceAttestationVendorReservedConstructor
+{
+public:
+    DeviceAttestationVendorReservedConstructor(struct VendorReservedElement * array, size_t size) : mElements(array), mMaxSize(size)
+    {}
+
+    typedef const struct VendorReservedElement * const_iterator;
+
+    const_iterator Next()
+    {
+        if ((mCurrentIndex + 1) == mNumEntriesUsed)
+        {
+            return nullptr;
+        }
+        return &mElements[++mCurrentIndex];
```

Comment:
This behavior is confusing, since the 0th element cannot be accessed by Next(), which is counter-intuitive since it means Next() must always be called at the bottom of the loop.

Suggest changing to postFix ++ and check for `if ((mCurrentIndex == mNumEntriesUsed)` at the beginning. Then it would be more symetrical and you could call Next().

You can leave as-is, but it would be more symetrical Next() giving you the next item and so that:

`while ((item = itr->Next()) != nullptr)` is a sufficient condition

Reply:
Addressed in https://github.com/project-chip/connectedhomeip/pull/10546/commits/421e3cd874cbd2c15cd31019e7358bca688f9303